accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,341 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+| late var x;       | Ok            | Ok                  | Ok              |
+| late final x;     | Error         | Ok                  | Ok              |
+| late int x;       | Ok            | Ok                  | Ok              |
+| late final int x; | Error         | Ok                  | Ok              |
+
+### Local variables and initialization based inference
+
+Local variables with explicitly written types are given the declared types as
+written.  The declared type of the variable is considered a "type of interest"
+in the sense defined in the flow analysis specification.  If the variable has an
+initializer (explicit or implicit) and is not `final`, then the declaration is
+treated as an assignment for the purposes of promotion.
+
+*Treating the declared type of the variable as a "type of interest" implies that
+if the variable is a nullable type, then its non-nullable version is also a type
+of interest.  Treating the initialization as an assignment for the purposes of
+promotion means that initializing a mutable variable declared at type `T?` with
+a value of non-nullable type `T` immediately promotes the variable to the
+non-nullable type.*
+
+```dart
+void test() {
+  int? x = 3; // x is declared at `int?`
+  x.isEven; // Valid, x has been promoted to `int`
+  x = null; // Valid, demotes to the declared type.
+}
+```
+
+Local variables with no explicitly written type but with an initializer are
+given an inferred declared type equal to the type of their initializer.  In the
+case that the type of the initializer is a promoted type variable `X & T`, the
+inferred type of the variable shall be `X`.  However, such a variable shall be
+treated as immediately promoted to `X & T`.
+
+#### Untyped variables
+
+Local variables with no explicitly written type and no initializer are subject
+to further promotion based inference.  The variables under consideration here
+are those declared as `var x;`, `final x;`, `late var x;`, or `late final x;`.
+We refer to such variables as **untyped variables**.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+that variable is in the **definitely unassigned** state.  Such an assignment on
+a path shall cause the variable in question to be treated by the flow analysis
+on that path as if it were declared with the assigned expression as its
+initializing expression, up until the next join point in the program.  We refer
+to this inferred type as the **initialization inferred type** of the variable.
+The **initialization inferred type** of the variable is always treated as a
+**type of interest** for the purposes of promotion (including its non-nullable
+version if applicable).
+
+```dart
+void test() {
+  var x; // x is an untyped variable
+  x = 3; // x is given the initialization inferred type of `int`
+}
+```
+
+At a join point in the program, it is an error if an **untyped variable** has
```

Comment:
I find your rewording very confusing, because the binding of the `non` part is ambiguous.  I think you mean "non-`late`, and `final`,", which is correct but awkward when written out that way, but as written it can also be read "non-('late` and `final`) which is incorrect.  

To your second question, no.  You can have variables that are definitely assigned but have no initialization inferred types.  That's one of the key points of all three of these proposals: that they separate the notion of having an initialization inferred type from being definitely assigned.

Reply:
English is weirdly ambiguous sometimes! In Danish I'd write `non-late and -final` if I wanted `non` to bind to both, and `non-late and final` if I didn't.

Apart from that, I like using "assigned" rather than having an inferred type. Then it's just a matter of permuting the conjunctive phrases until it is readable.

> If an untyped variable is `final` and non-`late`, it is an error if, at a join point, the variable is potentially assigned on at least one reachable path, and potentially unassigned on at least one other reachable path.

or

> It is an error if an untyped variable is declared `final` and not declared `late` and there are two reachable paths leading to a join point where the variable is potentially assigned on one path and potentially unassigned on the other.

or ...